perun-core/src/main/java/cz/metacentrum/perun/core/impl/UsersManagerImpl.java

```diff
@@ -794,90 +795,112 @@ public void removeAllUserExtSources(PerunSession sess, User user) {
+			} else {
+				throw new InternalErrorException("Unsupported db type");
+			}
+		} else {
+			// Part of query to search by user name (not exact)
+			if (Compatibility.isPostgreSql()) {
+				userNameQueryString = " strpos(lower(" + Compatibility.convertToAscii("COALESCE(users.first_name,'') || COALESCE(users.middle_name,'') || COALESCE(users.last_name,'')") + "),:nameString) > 0 ";
+			} else if (Compatibility.isHSQLDB()) {
+				userNameQueryString = " lower(" + Compatibility.convertToAscii("COALESCE(users.first_name,'') || COALESCE(users.middle_name,'') || COALESCE(users.last_name,'')") + ") like '%' || :nameString || '%' ";
+			} else {
+				throw new InternalErrorException("Unsupported db type");
+			}
+		}
 
-		// Search by mail (member)
-		users.addAll(jdbc.query("select " + userMappingSelectQuery +
-				" from users, members, member_attr_values, attr_names " +
-				"where members.user_id=users.id and members.id=member_attr_values.member_id and member_attr_values.attr_id=attr_names.id and " +
-				"attr_names.attr_name='urn:perun:member:attribute-def:def:mail' and " +
-				"lower(member_attr_values.attr_value)=lower(?)", USER_MAPPER, searchString));
-
-		// Search preferred email (user)
-		users.addAll(jdbc.query("select " + userMappingSelectQuery +
-				" from users, user_attr_values, attr_names " +
-				"where users.id=user_attr_values.user_id and user_attr_values.attr_id=attr_names.id and " +
-				"attr_names.attr_name='urn:perun:user:attribute-def:def:preferredMail' and " +
-				"lower(user_attr_values.attr_value)=lower(?)", USER_MAPPER, searchString));
-
-		// Search logins in userExtSources
-		users.addAll(jdbc.query("select " + userMappingSelectQuery +
-				" from users, user_ext_sources " +
-				"where lower(user_ext_sources.login_ext)=lower(?) and user_ext_sources.user_id=users.id", USER_MAPPER, searchString));
-
-		// Search logins in attributes: login-namespace:*
-		users.addAll(jdbc.query("select distinct " + userMappingSelectQuery +
-						" from attr_names, user_attr_values, users " +
-						"where attr_names.friendly_name like 'login-namespace:%' and lower(user_attr_values.attr_value)=lower(?) " +
-						"and attr_names.id=user_attr_values.attr_id and user_attr_values.user_id=users.id",
-				USER_MAPPER, searchString));
-
-		// Search by userId
+		// Part of query to search by user id
+		String idQueryString = "";
 		try {
-			int userId = Integer.parseInt(searchString);
-			users.addAll(jdbc.query("select " + userMappingSelectQuery + " from users where id=?", USER_MAPPER, userId));
+			int id = Integer.parseInt(searchString);
+			idQueryString = " users.id=" + id + " or ";
 		} catch (NumberFormatException e) {
-			// IGNORE
+			// IGNORE wrong format of ID
+		}
+
+		// Divide attributes received from CoreConfig into member, user and userExtSource attributes
+		List<String> allAttributes = BeansUtils.getCoreConfig().getAttributesToSearchUsersAndMembersBy();
```

Comment:
The next lines of code seem to be the same as in MembersManagerImpl. Could you extract it to a separate method that can be used in both managers?

Reply:
Ok, I've created new methods in Utils, although I'm not sure if it is not too complicated, I couldn't think of anything nicer.